.circleci/config.yml

```diff
@@ -30,13 +30,38 @@ jobs:
       - run: docker version
       - run: nproc
       - run: make clear-images
-      - run: make full-test -j $(nproc)
+
+      # Work out what needs rebuilding and what has changed
+      - run: |
+          ( ./changed-since-last.sh
+            echo "export DIND_CHANGED DIND_TAG DIND_PREV"
+            echo "export FN_CHANGED FN_TAG FN_PREV"
+          ) >> $BASH_ENV
+
+      # Rebuild the build tools if necessary
+      # This leaves local docker images around
+      - run: |
+          if [[ -n "$DIND_CHANGED" ]]; then
+            make build-dind
+          fi
+
+      # Rebuild fnserver if necessary
+      - run: |
+          if [[ -n "$FNLB_CHANGED" ]]; then
+            make full-test -j $(nproc)
+          fi
+
       - deploy:
           command: |
             if [[ "${CIRCLE_BRANCH}" == "master" && -z "${CIRCLE_PR_REPONAME}" ]]; then
               docker login -u $DOCKER_USER -p $DOCKER_PASS
               git config --global user.email "ci@fnproject.com"
               git config --global user.name "CI"
               git branch --set-upstream-to=origin/${CIRCLE_BRANCH} ${CIRCLE_BRANCH}
-              ./release.sh
+              if [[ -n "$DIND_CHANGED" ]]; then
+                make release-dind
+              fi
+              if [[ -n "$FN_CHANGED" ]]; then
```

Comment:
this should probably be every time (unconditional) too? if dind changes, so does fn, otherwise it's just fn changes in theory (or a bug in change checking).